lib/core/xskdev/xskdev.c

```diff
@@ -191,82 +191,50 @@ configure_busy_poll(xskdev_info_t *xi)
     return 0;
 }
 
-static int
-fq_reserved(xskdev_info_t *xi, uint16_t size)
+static inline void
+xsk_ring_prod__cancel(struct xsk_ring_prod *prod, __u32 nb)
 {
-    struct xskdev_umem *ux   = xi->rxq.ux;
-    struct xsk_ring_prod *fq = &ux->fq;
-    void *bufs[size];
-    uint16_t nb_bufs = size;
-    uint32_t pos     = 0;
-    int ret          = 0;
-
-    ret = xsk_ring_prod__reserve(fq, nb_bufs, &pos);
-    if (ret != nb_bufs)
-        return -1;
-
-    if (xskdev_buf_alloc(xi, bufs, nb_bufs) <= 0) {
-        xi->stats.fq_alloc_failed++;
-        return -1;
-    }
-    xi->stats.rx_buf_alloc += nb_bufs;
-
-    for (int i = 0; i < nb_bufs; i++) {
-        __u64 *fq_addr;
-        void *buf       = bufs[i];
-        uint64_t offset = (uint64_t)xskdev_buf_get_addr(xi, buf) - (uint64_t)xi->rxq.ux->umem_addr -
-                          (uint64_t)xi->buf_mgmt.pool_header_sz;
-
-        xskdev_buf_reset(xi, buf, xi->rxq.ux->obj_sz, xi->buf_mgmt.buf_headroom);
-
-        fq_addr  = xsk_ring_prod__fill_addr(fq, pos++);
-        *fq_addr = offset;
-    }
-
-    xsk_ring_prod__submit(fq, nb_bufs);
-
-    return 0;
+    prod->cached_prod -= nb;
 }
 
-static int
+static void
 fq_add(xskdev_info_t *xi)
 {
     struct xskdev_umem *ux   = xi->rxq.ux;
     struct xsk_ring_prod *fq = &ux->fq;
     void *bufs[FQ_ADD_BURST_COUNT];
-    uint16_t nb_bufs;
+    uint32_t nb_bufs;
     uint32_t pos = 0;
-    int nb;
 
-    nb_bufs = FQ_ADD_BURST_COUNT;
+    xi->stats.fq_add_called++;
 
-    if (xskdev_buf_alloc(xi, (void **)bufs, nb_bufs) <= 0) {
-        xi->stats.fq_alloc_failed++;
-        return -1;
-    }
-    xi->stats.rx_buf_alloc += nb_bufs;
-
-    nb = xsk_ring_prod__reserve(fq, nb_bufs, &pos);
+    for (;;) {
+        if (xsk_ring_prod__reserve(fq, FQ_ADD_BURST_COUNT, &pos) != FQ_ADD_BURST_COUNT) {
+            xi->stats.fq_reserve_failed++;
+            break;
+        }
 
-    for (int i = 0; i < nb; i++) {
-        void *buf       = bufs[i];
-        uint64_t offset = (uint64_t)xskdev_buf_get_addr(xi, buf) - (uint64_t)ux->umem_addr -
-                          (uint64_t)xi->buf_mgmt.pool_header_sz;
+        nb_bufs = xskdev_buf_alloc(xi, (void **)bufs, FQ_ADD_BURST_COUNT);
+        if (nb_bufs != FQ_ADD_BURST_COUNT) {
```

Comment:
So does this mean, if we try to allocate FQ_ADD_BURST_COUNT and we can't allocate all buffers  , cancel the buffers which got allocated by `xskdev_buf_alloc()`?  Is there a way to check the size of FQ to see if it can accommodate at least FQ_ADD_BURST_COUNT buffer before calling `xskdev_buf_alloc()` ?

Reply:
When we do the xsk_ring_prod__reserver() call we check to make sure we can add the buffers to the ring. Is this what you meant?

Reply:
ok. Does xsk_ring_prod__reserver() ensures we can allocate at least FQ_ADD_BURST_COUNT buffers ?

Reply:
The` xsk_ring_prod__reserve()` call above the allocation will insure we can put `FQ_ADD_BURST_COUNT` on the FQ ring. The `xskdev_buf_alloc()` call is checked to make sure we get `FQ_ADD_BURST_COUNT` of mbufs.